File: net.sourceforge.reb4j.scala/src/main/scala/net/sourceforge/reb4j/scala/charclass/CharRange.scala
package net.sourceforge.reb4j.scala.charclass

/**
 * Character class matching any character between a range of characters.
 */
final class CharRange private[charclass] (val first : Char, val last : Char) 
	extends CharClass 
	with WrappedNegation[CharRange] 
	with BracketsRequired 
	with Union.Subset 
	with Intersection.Superset
{
	require (first < last, "first must be < last")
	
	override lazy val unitableForm = first + "-" + last
	override def equals(other : Any) = other match
	{
		case that : CharRange => 
			this.first == that.first && this.last == that.last
		case _ => false
	}
	override lazy val hashCode = 
		31 * first.hashCode + last.hashCode
}

File: net.sourceforge.reb4j/src/main/java/net/sourceforge/reb4j/charclass/CharRange.java
package net.sourceforge.reb4j.charclass;

import fj.data.LazyString;

public final class CharRange extends CharClass
{
	private static final long serialVersionUID = 1L;
	public final char first, last;
	
	CharRange(final char first, final char last)
	{
		if (first >= last) throw new IllegalArgumentException("first must be < last");
		this.first = first;
		this.last = last;
	}

	@Override
	public Negated<CharRange> negated()
	{
		return new Negated<CharRange>(this);
	}

	@Override
	public LazyString unitableForm()
	{
		return LazyString.str(Character.toString(first)).append("-").append(Character.toString(last));
	}

	@Override
	public LazyString independentForm()
	{
		return LazyString.str("[").append(unitableForm()).append("]");
	}

}
